In [2]:
import pandas as pd
import numpy as np

### Load the Data

In [3]:
preprocessed_data = pd.read_csv('df_preprocessed.csv')
absenteeism_data = preprocessed_data.copy()
absenteeism_data.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day  \
0         0         0         0         1            7    7   
1         0         0         0         0            7   14   
2         0         0         0         1            7   15   
3         1         0         0         0            7   16   
4         0         0         0         1            7   23   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         2     0                          4  
4               30          0         2     1                          2

Since we are dealing with a logistic regression, we need to classify our targets into two classes and assign them values of 0 and 1. Here we will create two classes - Moderately absent and Excessively absent representing values zero and 1. In this dataset, we use the Day of the month rather than the day of the week to see the impact on Absenteeism

### Create the targets

In [4]:
# To classify our targets, we obtain the median value of the Absenteeism Time in Hours and then use it to create our
# two classes

np.median(absenteeism_data['Absenteeism Time in Hours'])

3.0

In [5]:
# We create a targets variable to hold the classified targets
targets = np.where(absenteeism_data['Absenteeism Time in Hours'] <= 
                   np.median(absenteeism_data['Absenteeism Time in Hours']), 0, 1)
targets

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [6]:
absenteeism_data['Excessive Absenteeism'] = targets
absenteeism_data

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day  \
0           0         0         0         1            7    7   
1           0         0         0         0            7   14   
2           0         0         0         1            7   15   
3           1         0         0         0            7   16   
4           0         0         0         1            7   23   
..        ...       ...       ...       ...          ...  ...   
695         1         0         0         0            5   23   
696         1         0         0         0            5   23   
697         1         0         0         0            5   24   
698         0         0         0         1            5   24   
699         0         0         0         1            5   31   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0                 30          0         2     1                          4   
1                 31          0         1     0                          0   
2                 31          0         0     0                          2   
3                 24          0         2     0                          4   
4                 30          0         2     1                          2   
..               ...        ...       ...   ...                        ...   
695               22          1         2     0                          8   
696               24          0         1     2                          3   
697               25          1         0     0                          8   
698               25          1         0     0                          2   
699               25          0         1     1                          2   

     Excessive Absenteeism  
0                        1  
1                        0  
2                        0  
3                        1  
4                        0  
..                     ...  
695                      1  
696                      0  
697                      1  
698                      0  
699                      0  

[700 rows x 16 columns]

In [7]:
# Using the median for mapping our targets has also helped us accomplish another goal implicitly - Balancing the 
# dataset. This allows our priors to be balances such that we do not train or fit our model on unbalanced priors.
# To validate this, we take the sum of the targets and divide by the total number of targets

priors = targets.sum()/targets.shape[0]
priors

0.45571428571428574

In [8]:
# We drop the Absenteeism in Hours column as it is no longer needed

data_with_targets = absenteeism_data.drop(['Absenteeism Time in Hours'], axis = 1)

In [9]:
data_with_targets.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day  \
0         0         0         0         1            7    7   
1         0         0         0         0            7   14   
2         0         0         0         1            7   15   
3         1         0         0         0            7   16   
4         0         0         0         1            7   23   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Excessive Absenteeism  
0               30          0         2     1                      1  
1               31          0         1     0                      0  
2               31          0         0     0                      0  
3               24          0         2     0                      1  
4               30          0         2     1                      0

In [10]:
# We specify our input variables
unscaled_inputs = data_with_targets.iloc[:,:-1]

### Scaling the inputs

In [11]:
from sklearn.preprocessing import StandardScaler

In [12]:
scaler = StandardScaler() # We create a scaler object here in order to scale some parts of the inputs as we do not
# want to scale the dummies

scaler.fit(unscaled_inputs[['Month Value','Day', 'Transportation Expense', 'Distance to Work',
                           'Age', 'Daily Work Load Average', 'Body Mass Index', 'Children','Pets']])

StandardScaler(copy=True, with_mean=True, with_std=True)

In [13]:
unscaled_inputs[['Month Value','Day', 'Transportation Expense', 'Distance to Work',
                           'Age', 'Daily Work Load Average', 'Body Mass Index', 'Children','Pets']] = scaler.transform(
unscaled_inputs[['Month Value','Day', 'Transportation Expense', 'Distance to Work',
                           'Age', 'Daily Work Load Average', 'Body Mass Index', 'Children','Pets']])

In [14]:
scaled_inputs = np.array(unscaled_inputs)

In [15]:
scaled_inputs

array([[ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.88046927,  0.26848661],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
        -0.01928035, -0.58968976],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
        -0.91902997, -0.58968976],
       ...,
       [ 1.        ,  0.        ,  0.        , ...,  1.        ,
        -0.91902997, -0.58968976],
       [ 0.        ,  0.        ,  0.        , ...,  1.        ,
        -0.91902997, -0.58968976],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
        -0.01928035,  0.26848661]])

In [16]:
scaled_inputs.shape

(700, 14)

### Define the target

In [17]:
targets = data_with_targets.iloc[:,-1]

In [18]:
targets

0      1
1      0
2      0
3      1
4      0
      ..
695    1
696    0
697    1
698    0
699    0
Name: Excessive Absenteeism, Length: 700, dtype: int64

### Shuffle the dataset and split

In [19]:
# shuffled_indices = np.arange(scaled_inputs.shape[0])
# np.random.shuffle(shuffled_indices)

# shuffled_inputs = scaled_inputs[shuffled_indices]
# shuffled_targets = targets[shuffled_indices]

In [20]:
# shuffled_inputs.shape

In [21]:
# shuffled_targets.shape

In [22]:
# Split the dataset into training and testing data. We use an 80-20 split for the dataset. The shuffle
# parameter is set to true by default to shuffle the dataset before splitting. To also ensure this dataset 
# is not shuffled everytime we re-run the code. We use a random state of 365

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(scaled_inputs, targets,
                                                    test_size = 0.2, random_state=365)

In [23]:
x_train.shape, x_test.shape

((560, 14), (140, 14))

In [24]:
y_train.shape, y_test.shape

((560,), (140,))

### Model Training

In [25]:
# import statsmodels.api as sm

In [26]:
# x = sm.add_constant(x_train)
# log_result = sm.Logit(y_train,x).fit()

In [27]:
# log_result.summary()

In [28]:
# From the output above, we see that statsmodels runs out of iterations during model training. Hence we need to use 
# another library to train the model

from sklearn.linear_model import LogisticRegression

In [29]:
log_model = LogisticRegression()
log_model.fit(x_train,y_train)

/usr/local/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [30]:
# To check the performance of the model i.e how well our model learn how to classify inputs based on the training data
# we use the score(R-squared)

log_model.score(x_train,y_train)

0.775

In [31]:
# From the score, we see that our model learns how to classify up to 76% of the data accurately, now let us check that
# manually to verify

### Checking the Model accuracy

In [32]:
# We bascially compare the outputs with the targets. The outputs are predictions of the training inputs
model_output = log_model.predict(x_train)
model_output

array([1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1,
       1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1,
       0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1,
       0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0,

In [33]:
np.array(y_train)

array([1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1,
       1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1,
       0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0,

In [34]:
model_output == np.array(y_train)

array([ True,  True, False,  True,  True,  True, False,  True, False,
        True,  True,  True,  True, False,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True, False,
       False, False,  True,  True,  True,  True,  True,  True, False,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
       False, False,  True, False,  True,  True,  True,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True, False,  True, False,
        True,  True,  True,  True, False,  True,  True, False, False,
       False,  True, False,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True,  True,  True,
        True, False,  True,  True,  True, False, False,  True, False,
       False, False,  True,  True, False,  True,  True,  True,  True,
        True,  True,

In [35]:
# Since these boolean values are represented as 0s and 1s, we run a sum over the comparison and divide by the number
# of observations

model_accuracy = np.sum(model_output == np.array(y_train))/x_train.shape[0]
model_accuracy

0.775

In [36]:
# Fetch the coefficients and intercept
log_model.coef_

array([[ 2.50791017,  1.28147188,  2.78528033,  0.67995342,  0.13311093,
        -0.06763024,  0.52040999, -0.056014  , -0.26317466, -0.03303039,
         0.32677148, -0.16981508,  0.42835123, -0.30116593]])

In [37]:
log_model.intercept_

array([-1.36411165])

### Create a Summary Table

In [38]:
# We create a summary table to hold the features and their coefficients, as well as the intercept. 
#Since sklearn makes use of arrays, we will have to transform our data into dataframes

features = unscaled_inputs.columns.values
summary_table = pd.DataFrame(columns=['Features'], data=features)
summary_table['coefficients'] = np.transpose(log_model.coef_) # By default, ndarrays are rows only so we transpose them
# summary_table['intercept'] = np.transpose(log_model.intercept_)

summary_table

Features  coefficients
0                  Reason_1      2.507910
1                  Reason_2      1.281472
2                  Reason_3      2.785280
3                  Reason_4      0.679953
4               Month Value      0.133111
5                       Day     -0.067630
6    Transportation Expense      0.520410
7          Distance to Work     -0.056014
8                       Age     -0.263175
9   Daily Work Load Average     -0.033030
10          Body Mass Index      0.326771
11                Education     -0.169815
12                 Children      0.428351
13                     Pets     -0.301166

In [39]:
summary_table.index = summary_table.index + 1
summary_table.loc[0] = ['Intercept', log_model.intercept_[0]]
summary_table = summary_table.sort_index()
summary_table

Features  coefficients
0                 Intercept     -1.364112
1                  Reason_1      2.507910
2                  Reason_2      1.281472
3                  Reason_3      2.785280
4                  Reason_4      0.679953
5               Month Value      0.133111
6                       Day     -0.067630
7    Transportation Expense      0.520410
8          Distance to Work     -0.056014
9                       Age     -0.263175
10  Daily Work Load Average     -0.033030
11          Body Mass Index      0.326771
12                Education     -0.169815
13                 Children      0.428351
14                     Pets     -0.301166

In [40]:
# Since logistic deals with the log(odds), we can calculate the odds ratio of the coefficients to determine which 
# coefficients have the highest weights

summary_table['odds_ratio'] = np.exp(summary_table.coefficients)
summary_table

Features  coefficients  odds_ratio
0                 Intercept     -1.364112    0.255608
1                  Reason_1      2.507910   12.279242
2                  Reason_2      1.281472    3.601937
3                  Reason_3      2.785280   16.204360
4                  Reason_4      0.679953    1.973786
5               Month Value      0.133111    1.142377
6                       Day     -0.067630    0.934606
7    Transportation Expense      0.520410    1.682717
8          Distance to Work     -0.056014    0.945526
9                       Age     -0.263175    0.768608
10  Daily Work Load Average     -0.033030    0.967509
11          Body Mass Index      0.326771    1.386485
12                Education     -0.169815    0.843821
13                 Children      0.428351    1.534725
14                     Pets     -0.301166    0.739955

In [41]:
summary_table = summary_table.sort_values(['odds_ratio'], ascending=False)
summary_table

Features  coefficients  odds_ratio
3                  Reason_3      2.785280   16.204360
1                  Reason_1      2.507910   12.279242
2                  Reason_2      1.281472    3.601937
4                  Reason_4      0.679953    1.973786
7    Transportation Expense      0.520410    1.682717
13                 Children      0.428351    1.534725
11          Body Mass Index      0.326771    1.386485
5               Month Value      0.133111    1.142377
10  Daily Work Load Average     -0.033030    0.967509
8          Distance to Work     -0.056014    0.945526
6                       Day     -0.067630    0.934606
12                Education     -0.169815    0.843821
9                       Age     -0.263175    0.768608
14                     Pets     -0.301166    0.739955
0                 Intercept     -1.364112    0.255608

### Interpreting the Summary table

We know that a feature is not particularly important when the coefficient is zero or the odds ratio is 1 i.e. for 
each unit increase in that feature, the odds will increase by a multiple of the odds ratio. Therefore, a ratio of 1
gives us no increase in the odds.

The features that have the most impact on excessive seem to be each of the reasons given with each group as follows:
1. Various Diseases
2. Pregnancy and childbirth
3. Poisoning
4. Light diseases

We see that an employee is 15 times more likely to be excessively absent when they are poisoned, which seems reasonable, and 12, 3 and 2 for reasons 1, 2 and 4 respectively

The transportation expense is next on the hierarchy as we see that for each unit increase in Transportation expense, the employee is more than 1.5 times likely to be excessively absent

The pets also informs us that for each(standardized) unit increase in the number of pets, an employee is less likely
to be absent. This could infer that with an increase in the number of pets, an employee could hire a caretaker rather than being absent from work OR employees with pets tend to be happier and are therefore less absent at work.

The Education tells us that for employees with a higher level of education, they are less likely to be excessively absent from work. 

In another notebook, we simplify this model by dropping the features with low weights(coefficients). These include the following
1. Month Value
2. Daily Work Load Average
3. Distance to work 

These are the features with weights most closest to zero. 